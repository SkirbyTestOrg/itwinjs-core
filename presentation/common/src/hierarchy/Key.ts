/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Hierarchies */

import { InstanceKey, InstanceKeyJSON } from "../EC";

/**
 * Standard node types
 * @public
 */
export enum StandardNodeTypes {
  ECInstanceNode = "ECInstanceNode",
  ECClassGroupingNode = "ECClassGroupingNode",
  ECPropertyGroupingNode = "ECPropertyGroupingNode",
  DisplayLabelGroupingNode = "DisplayLabelGroupingNode",
}

/**
 * One of the node key types
 * @public
 */
export type NodeKey = BaseNodeKey | ECInstanceNodeKey | ECClassGroupingNodeKey | ECPropertyGroupingNodeKey | LabelGroupingNodeKey;
/** @public */
export namespace NodeKey {
  /**
   * Serialize given [[NodeKey]] to JSON
   * @internal
   */
  export function toJSON(key: NodeKey): NodeKeyJSON {
    if (isInstanceNodeKey(key))
      return { ...key, instanceKey: InstanceKey.toJSON(key.instanceKey) };
    return { ...key };
  }

  /**
   * Deserialize node key from JSON
   * @param json JSON or JSON serialized to string to deserialize from
   * @returns Deserialized node key
   *
   * @internal
   */
  export function fromJSON(json: NodeKeyJSON): NodeKey {
    if (isInstanceNodeKey(json))
      return { ...json, instanceKey: InstanceKey.fromJSON(json.instanceKey) };
    return { ...json };
  }

  /**
   * Checks if the supplied key is an [[ECInstanceNodeKey]]
   * @public
   */
  export function isInstanceNodeKey(key: NodeKey): key is ECInstanceNodeKey {
    return key.type === StandardNodeTypes.ECInstanceNode;
  }

  /**
   * Checks if the supplied key is an [[ECClassGroupingNodeKey]]
   * @public
   */
  export function isClassGroupingNodeKey(key: NodeKey): key is ECClassGroupingNodeKey {
    return key.type === StandardNodeTypes.ECClassGroupingNode;
  }

  /**
   * Checks if the supplied key is an [[ECPropertyGroupingNodeKey]]
   * @public
   */
  export function isPropertyGroupingNodeKey(key: NodeKey): key is ECPropertyGroupingNodeKey {
    return key.type === StandardNodeTypes.ECPropertyGroupingNode;
  }

  /**
   * Checks if the supplied key is a [[LabelGroupingNodeKey]]
   * @public
   */
  export function isLabelGroupingNodeKey(key: NodeKey): key is LabelGroupingNodeKey {
    return key.type === StandardNodeTypes.DisplayLabelGroupingNode;
  }

  /**
   * Checks if the supplied key is a grouping node key
   * @public
   */
  export function isGroupingNodeKey(key: NodeKey): key is GroupingNodeKey {
    return isClassGroupingNodeKey(key) || isPropertyGroupingNodeKey(key) || isLabelGroupingNodeKey(key);
  }
}

/**
 * Node key path. Can be used to define path from one node to another.
 * @public
 */
export type NodeKeyPath = NodeKey[];

/**
 * Data structure that describes a basic node key
 * @public
 */
export interface BaseNodeKey {
  /** Node type */
  type: string;
  /** Node hash path from root to the node whose key this is */
  pathFromRoot: string[];
}

/**
 * Data structure that describes an ECInstance node key
 * @public
 */
export interface ECInstanceNodeKey extends BaseNodeKey {
  type: StandardNodeTypes.ECInstanceNode;
  /** [[InstanceKey]] of the ECInstance represented by the node */
  instanceKey: InstanceKey;
}

/**
 * Serialized [[ECInstanceNodeKey]] JSON representation.
 * @internal
 */
export interface ECInstanceNodeKeyJSON extends BaseNodeKey {
  type: StandardNodeTypes.ECInstanceNode;
  instanceKey: InstanceKeyJSON;
}

/**
 * Data structure that describes a grouping node key
 * @public
 */
export interface GroupingNodeKey extends BaseNodeKey {
  /**
   * Get the number of instances grouped by the node represented
   * by this key.
   *
   * **Note:** this property is just a helper and is not involved
   * in identifying a node.
   */
  groupedInstancesCount: number;
}

/**
 * Data structure that describes an ECClass grouping node key
 * @public
 */
export interface ECClassGroupingNodeKey extends GroupingNodeKey {
  type: StandardNodeTypes.ECClassGroupingNode;
  /** Full name of the grouping ECClass */
  className: string;
}

/**
 * Data structure that describes an ECProperty grouping node key
 * @public
 */
export interface ECPropertyGroupingNodeKey extends GroupingNodeKey {
  type: StandardNodeTypes.ECPropertyGroupingNode;
  /** Full name of the grouping ECProperty class */
  className: string;
  /** Name of the ECProperty */
  propertyName: string;
  /** Grouping value */
  groupingValue: any;
}

/**
 * Data structure that describes a display label grouping node key
 * @public
 */
export interface LabelGroupingNodeKey extends GroupingNodeKey {
  type: StandardNodeTypes.DisplayLabelGroupingNode;
  /** Grouping display label */
  label: string;
}

/**
 * One of the serialized node key types
 * @internal
 */
export type NodeKeyJSON = BaseNodeKey | ECInstanceNodeKeyJSON | ECClassGroupingNodeKey | ECPropertyGroupingNodeKey | LabelGroupingNodeKey;
