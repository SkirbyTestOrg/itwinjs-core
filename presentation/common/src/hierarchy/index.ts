/*---------------------------------------------------------------------------------------------
* Copyright (c) 2018 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Hierarchies */

export { NodeKey, NodeKeyPath, StandardNodeTypes } from "./Key";
export { BaseNodeKey, ECInstanceNodeKey, ECClassGroupingNodeKey, ECPropertyGroupingNodeKey, LabelGroupingNodeKey } from "./Key";
export { default as Node } from "./Node";
export { default as NodePathElement } from "./NodePathElement";
