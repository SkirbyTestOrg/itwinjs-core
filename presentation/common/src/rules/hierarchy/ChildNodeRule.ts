/*---------------------------------------------------------------------------------------------
* Copyright (c) 2018 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module PresentationRules */

import { NavigationRuleBase } from "./NavigationRule";
import { RuleTypes, ConditionContainer } from "../Rule";

/**
 * Child node rules define nodes that are displayed at
 * each child hierarchy level.
 */
export interface ChildNodeRule extends NavigationRuleBase, ConditionContainer {
  /** Used for serializing to JSON. */
  ruleType: RuleTypes.ChildNodes;

  /**
   * Defines a condition for the rule, which needs to be met in order to execute it. Condition
   * is an [ECExpression]($docs/learning/ECExpressions.md), which can use
   * a [limited set of symbols]($docs/learning/hierarchies/ECExpressions.md#rule-condition).
   */
  condition?: string;
}
