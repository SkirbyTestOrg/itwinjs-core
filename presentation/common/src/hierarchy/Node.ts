/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Hierarchies */

import { NodeKey, NodeKeyJSON, nodeKeyFromJSON } from "./Key";

/**
 * Data structure that describes a tree node.
 */
export default interface Node {
  /** A key that uniquely identifies a node. */
  key: NodeKey;
  /** Display label */
  label: string;
  /** Extensive description */
  description?: string;
  /** Image ID */
  imageId?: string;
  /** Foreground color */
  foreColor?: string;
  /** Background color */
  backColor?: string;
  /** Font style */
  fontStyle?: string;
  /** Does this node have child nodes */
  hasChildren?: boolean;
  /** Is this node selectable */
  isSelectionDisabled?: boolean;
  /** Is this node editable */
  isEditable?: boolean;
  /** Is this node expanded */
  isExpanded?: boolean;
  /** Is checkbox visible for this node */
  isCheckboxVisible?: boolean;
  /** Is this node's checkbox checked */
  isChecked?: boolean;
  /** Is this node's checkbox enabled */
  isCheckboxEnabled?: boolean;
}

/**
 * Serialized [[Node]] JSON representation.
 *
 * @hidden
 */
export interface NodeJSON {
  key: NodeKeyJSON;
  label: string;
  description?: string;
  imageId?: string;
  foreColor?: string;
  backColor?: string;
  fontStyle?: string;
  hasChildren?: boolean;
  isSelectionDisabled?: boolean;
  isEditable?: boolean;
  isExpanded?: boolean;
  isCheckboxVisible?: boolean;
  isChecked?: boolean;
  isCheckboxEnabled?: boolean;
}

/**
 * Deserialize node from JSON
 * @param json JSON or JSON serialized to string to deserialize from
 * @returns Deserialized node
 *
 * @hidden
 */
export const fromJSON = (json: NodeJSON | string): Node => {
  if (typeof json === "string")
    return JSON.parse(json, reviver);
  return Object.assign({}, json, {
    key: nodeKeyFromJSON(json.key),
  });
};

/**
 * Reviver function that can be used as a second argument for
 * `JSON.parse` method when parsing [[Node]] objects.
 *
 * @hidden
 */
export const reviver = (key: string, value: any): any => {
  return key === "" ? fromJSON(value) : value;
};

/**
 * Deserialize nodes list from JSON
 * @param json JSON or JSON serialized to string to deserialize from
 * @returns Deserialized nodes list
 *
 * @hidden
 */
export const listFromJSON = (json: NodeJSON[] | string): Node[] => {
  if (typeof json === "string")
    return JSON.parse(json, listReviver);
  return json.map((m) => fromJSON(m));
};

/**
 * Reviver function that can be used as a second argument for
 * `JSON.parse` method when parsing [[Node]][] objects.
 *
 * @hidden
 */
export const listReviver = (key: string, value: any): any => {
  return key === "" ? listFromJSON(value) : value;
};
