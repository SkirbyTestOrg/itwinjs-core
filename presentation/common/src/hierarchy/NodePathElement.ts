/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Hierarchies */

import { default as Node, NodeJSON, fromJSON as nodeFromJson } from "./Node";

/** Data related to node hierachy filtering */
export interface NodePathFilteringData {
  /** FilterText occurances in a node label */
  occurances: number;
  /** FilterText occurances in all of node children */
  childrenOccurances: number;
}

/**
 * Describes a single step in the nodes path.
 */
export default interface NodePathElement {
  node: Node;
  index: number;
  isMarked: boolean;
  children: NodePathElement[];
  filteringData?: NodePathFilteringData;
}

/**
 * Serialized [[NodePathElement]] JSON representation.
 *
 * @hidden
 */
export interface NodePathElementJSON {
  node: NodeJSON;
  index: number;
  isMarked: boolean;
  children: NodePathElementJSON[];
  filteringData?: NodePathFilteringData;
}

/**
 * Deserialize [[NodePathElement]] from JSON
 * @param json JSON or JSON serialized to string to deserialize from
 * @returns Deserialized [[NodePathElement]]
 *
 * @hidden
 */
export const fromJSON = (json: NodePathElementJSON | string): NodePathElement => {
  if (typeof json === "string")
    return JSON.parse(json, reviver);
  return Object.assign({}, json, {
    node: nodeFromJson(json.node),
    children: listFromJSON(json.children),
  });
};

/**
 * Reviver function that can be used as a second argument for
 * `JSON.parse` method when parsing [[NodePathElement]] objects.
 *
 * @hidden
 */
export const reviver = (key: string, value: any): any => {
  return key === "" ? fromJSON(value) : value;
};

/**
 * Deserialize [[NodePathElement]] list from JSON
 * @param json JSON or JSON serialized to string to deserialize from
 * @returns Deserialized [[NodePathElement]] list
 *
 * @hidden
 */
export const listFromJSON = (json: NodePathElementJSON[] | string): NodePathElement[] => {
  if (typeof json === "string")
    return JSON.parse(json, listReviver);
  return json.map((m) => fromJSON(m));
};

/**
 * Reviver function that can be used as a second argument for
 * `JSON.parse` method when parsing [[NodePathElement]][] objects.
 *
 * @hidden
 */
export const listReviver = (key: string, value: any): any => {
  return key === "" ? listFromJSON(value) : value;
};
