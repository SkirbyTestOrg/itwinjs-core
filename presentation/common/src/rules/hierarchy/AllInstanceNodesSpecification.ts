/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module PresentationRules */
import { ChildNodeSpecificationBase, ChildNodeSpecificationTypes, DefaultGroupingPropertiesContainer } from "./ChildNodeSpecification";
import { SchemasSpecification } from "../SchemasSpecification";

/**
 * Creates nodes for all available instances filtered only by the
 * [[supportedSchemas]] properties of the specification or the ruleset.
 *
 * @public
 */
export interface AllInstanceNodesSpecification extends ChildNodeSpecificationBase, DefaultGroupingPropertiesContainer {
  /** Used for serializing to JSON. */
  specType: ChildNodeSpecificationTypes.AllInstanceNodes;

  /**
   * Specification of schemas whose instances should be returned.
   */
  supportedSchemas?: SchemasSpecification;
}
