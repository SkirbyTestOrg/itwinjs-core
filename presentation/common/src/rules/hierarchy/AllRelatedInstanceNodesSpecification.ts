/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module PresentationRules */

import { ChildNodeSpecificationBase, DefaultGroupingPropertiesContainer } from "./ChildNodeSpecification";
import { RuleSpecificationTypes } from "../RuleSpecification";
import { RelationshipDirection } from "../RelationshipDirection";
import { SchemasSpecification } from "../SchemasSpecification";

/**
 * Returns all related instance nodes for parent ECInstance node.
 *
 * **Precondition:** can be used only if parent node is ECInstance node. If there is
 * no immediate parent instance node it will go up until it finds one.
 */
export interface AllRelatedInstanceNodesSpecification extends ChildNodeSpecificationBase, DefaultGroupingPropertiesContainer {
  /** Used for serializing to JSON. */
  specType: RuleSpecificationTypes.AllRelatedInstanceNodes;

  /**
   * Specification of schemas whose instances should be returned.
   */
  supportedSchemas?: SchemasSpecification;

  /**
   * Direction that will be followed in the relationship select criteria.
   * Defaults to [[RelationshipDirection.Both]].
   */
  requiredDirection?: RelationshipDirection;

  /**
   * Skips defined level of related items and shows next level related items.
   *
   * **Example:** lets say we have relationship hierarchy A->B->C, but we want
   * to show A->C. In this case we need to set this property to 1 - the engine will
   * query all direct relationships and all nested relationships and only nested
   * ones will be used to create content.
   *
   * Defaults to `0`.
   *
   * @type integer
   * @minimum 0
   */
  skipRelatedLevel?: number;
}
