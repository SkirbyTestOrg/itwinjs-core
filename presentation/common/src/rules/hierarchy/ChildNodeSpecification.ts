/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module PresentationRules */

import { ChildNodeRule } from "./ChildNodeRule";
import { AllInstanceNodesSpecification } from "./AllInstanceNodesSpecification";
import { AllRelatedInstanceNodesSpecification } from "./AllRelatedInstanceNodesSpecification";
import { CustomNodeSpecification } from "./CustomNodeSpecification";
import { InstanceNodesOfSpecificClassesSpecification } from "./InstanceNodesOfSpecificClassesSpecification";
import { RelatedInstanceNodesSpecification } from "./RelatedInstanceNodesSpecification";
import { CustomQueryInstanceNodesSpecification } from "./CustomQueryInstanceNodesSpecification";
import { RelatedInstanceSpecification } from "../RelatedInstanceSpecification";

/**
 * Used for serializing array of [[ChildNodeSpecification]]
 * @public
 */
export enum ChildNodeSpecificationTypes {
  // hierarchy specifications
  AllInstanceNodes = "AllInstanceNodes",
  AllRelatedInstanceNodes = "AllRelatedInstanceNodes",
  RelatedInstanceNodes = "RelatedInstanceNodes",
  InstanceNodesOfSpecificClasses = "InstanceNodesOfSpecificClasses",
  CustomQueryInstanceNodes = "CustomQueryInstanceNodes",
  CustomNode = "CustomNode",
}

/**
 * Base interface for all [[ChildNodeSpecification]] implementations. Not
 * meant to be used directly, see `ChildNodeSpecification`.
 *
 * @public
 */
export interface ChildNodeSpecificationBase {
  /** Used for serializing to JSON. */
  specType: ChildNodeSpecificationTypes;

  /**
   * Defines the order in which specifications are evaluated and executed. Defaults to `1000`.
   *
   * @type integer
   */
  priority?: number;

  /**
   * This tells the rules engine that nodes produced using this
   * specification always or never have children. Defaults to `Unknown`.
   *
   * **Note:** setting this flag to `Always` or `Never` improves performance.
   */
  hasChildren?: "Always" | "Never" | "Unknown";

  /**
   * Hide nodes provided by this specification and directly show nodes of its children.
   * This helps if you need to define related instance nodes of particular parent node that is not available in the
   * hierarchy.
   */
  hideNodesInHierarchy?: boolean;

  /**
   * Hide nodes if they don't have children.
   */
  hideIfNoChildren?: boolean;

  /**
   * Set this flag to `true` to suppress default sorting of ECInstances returned by this specification.
   *
   * **Note:** setting this flag to `true` improves performance.
   */
  doNotSort?: boolean;

  /** Specifications of related instances that can be used in content creation. */
  relatedInstances?: RelatedInstanceSpecification[];

  /** [Nested rule]($docs/learning/hierarchies/Terminology.md#nested-rules) specifications. */
  nestedRules?: ChildNodeRule[];
}

/**
 * A container of default grouping properties. Used for specifications that support
 * default grouping. Not meant to be used directly, see `ChildNodeSpecification`.
 *
 * @public
 */
export interface DefaultGroupingPropertiesContainer {
  /** Group instances by ECClass. Defaults to `true`. */
  groupByClass?: boolean;

  /** Group instances by label. Defaults to `true`. */
  groupByLabel?: boolean;
}

/**
 * Navigation rule specifications that define what content the rule results in.
 * @public
 */
export type ChildNodeSpecification = AllInstanceNodesSpecification |
  AllRelatedInstanceNodesSpecification |
  CustomNodeSpecification |
  InstanceNodesOfSpecificClassesSpecification |
  RelatedInstanceNodesSpecification |
  CustomQueryInstanceNodesSpecification;
