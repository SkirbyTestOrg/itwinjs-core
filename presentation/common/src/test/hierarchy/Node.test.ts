/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
import { expect } from "chai";
import { Node, NodeJSON } from "../../hierarchy/Node";
import {
  createRandomECInstanceNode, createRandomECInstanceNodeKey,
  createRandomECInstanceKeyJSON,
} from "../_helpers/random";
import { ECInstanceNodeKeyJSON } from "../../hierarchy/Key";

const createRandomECInstanceNodeKeyJSON = (): ECInstanceNodeKeyJSON => {
  return {
    ...createRandomECInstanceNodeKey(),
    instanceKey: createRandomECInstanceKeyJSON(),
  };
};

const createRandomNodeJSON = (): NodeJSON => {
  return {
    ...createRandomECInstanceNode(),
    key: createRandomECInstanceNodeKeyJSON(),
  };
};

describe("Node", () => {

  describe("toJSON", () => {

    it("serializes Node", () => {
      const node = createRandomECInstanceNode();
      const json = Node.toJSON(node);
      expect(json).to.matchSnapshot();
    });

  });

  describe("fromJSON", () => {

    it("creates valid Node from JSON", () => {
      const json = createRandomNodeJSON();
      const node = Node.fromJSON(json);
      expect(node).to.matchSnapshot();
    });

    it("creates valid Node from serialized JSON", () => {
      const json = createRandomNodeJSON();
      const node = Node.fromJSON(JSON.stringify(json));
      expect(node).to.matchSnapshot();
    });

  });

  describe("listFromJSON", () => {

    it("creates valid Node[] from JSON", () => {
      const json = [createRandomNodeJSON(), createRandomNodeJSON()];
      const nodes = Node.listFromJSON(json);
      expect(nodes).to.matchSnapshot();
    });

    it("creates valid Node[] from serialized JSON", () => {
      const json = [createRandomNodeJSON(), createRandomNodeJSON()];
      const nodes = Node.listFromJSON(JSON.stringify(json));
      expect(nodes).to.matchSnapshot();
    });

  });

});
