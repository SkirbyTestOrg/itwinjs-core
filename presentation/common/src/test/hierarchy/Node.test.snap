// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Node fromJSON creates valid Node from JSON 1`] = `
Object {
  "backColor": undefined,
  "description": undefined,
  "foreColor": "#80E076",
  "hasChildren": true,
  "imageId": "customized",
  "isCheckboxEnabled": true,
  "isCheckboxVisible": false,
  "isChecked": true,
  "isEditable": true,
  "isExpanded": true,
  "isSelectionDisabled": false,
  "key": Object {
    "instanceKey": Object {
      "className": "Functionality",
      "id": "0x154f7000000a54e",
    },
    "pathFromRoot": Array [
      "4843c512-9f82-4b52-b6d8-74ba3b8ff740",
      "4881193c-08eb-445a-b7dd-b7a30f33c975",
    ],
    "type": "ECInstanceNode",
  },
  "label": "Technician Shoes",
}
`;

exports[`Node fromJSON creates valid Node from serialized JSON 1`] = `
Object {
  "backColor": "rgb(
    62,
    40,
    232
  )",
  "description": "Rerum eius fugiat dolores repellat ut excepturi.",
  "hasChildren": false,
  "isCheckboxEnabled": true,
  "isCheckboxVisible": true,
  "isChecked": false,
  "isEditable": false,
  "isExpanded": false,
  "isSelectionDisabled": false,
  "key": Object {
    "instanceKey": Object {
      "className": "website",
      "id": "0x5e58000000b378",
    },
    "pathFromRoot": Array [
      "5242abda-2a79-46dd-b8d9-fe71dab8e435",
      "8d78e06d-c620-4f5a-82b9-ecb87afe9470",
    ],
    "type": "ECInstanceNode",
  },
  "label": "Tennessee THX",
}
`;

exports[`Node listFromJSON creates valid Node[] from JSON 1`] = `
Array [
  Object {
    "backColor": "rgb(
    8,
    226,
    135
  )",
    "description": "Officia id quia ducimus porro ipsum ut labore non odit.",
    "foreColor": undefined,
    "hasChildren": true,
    "imageId": "payment",
    "isCheckboxEnabled": true,
    "isCheckboxVisible": false,
    "isChecked": false,
    "isEditable": true,
    "isExpanded": true,
    "isSelectionDisabled": true,
    "key": Object {
      "instanceKey": Object {
        "className": "generating",
        "id": "0xa1f40000015019",
      },
      "pathFromRoot": Array [
        "a4e3157e-9efc-44ab-9ccd-51f6568818bb",
        "5969c069-8a07-4d76-8acd-5b9e0d056fcd",
      ],
      "type": "ECInstanceNode",
    },
    "label": "Massachusetts",
  },
  Object {
    "backColor": undefined,
    "description": undefined,
    "foreColor": undefined,
    "hasChildren": true,
    "imageId": "calculate",
    "isCheckboxEnabled": false,
    "isCheckboxVisible": true,
    "isChecked": false,
    "isEditable": true,
    "isExpanded": true,
    "isSelectionDisabled": true,
    "key": Object {
      "instanceKey": Object {
        "className": "algorithm",
        "id": "0x116990000004b3f",
      },
      "pathFromRoot": Array [
        "e690a051-8d92-4dd1-89f9-a8a212f6872f",
        "d8f9f659-cef9-4c2d-bb44-acc219c15dbf",
      ],
      "type": "ECInstanceNode",
    },
    "label": "compress payment Tanzanian Shilling",
  },
]
`;

exports[`Node listFromJSON creates valid Node[] from serialized JSON 1`] = `
Array [
  Object {
    "backColor": "rgb(
    142,
    32,
    151
  )",
    "description": "Qui facilis nihil veniam non quia et odit.",
    "foreColor": "#825EAE",
    "hasChildren": true,
    "isCheckboxEnabled": true,
    "isCheckboxVisible": true,
    "isChecked": true,
    "isEditable": true,
    "isExpanded": true,
    "isSelectionDisabled": true,
    "key": Object {
      "instanceKey": Object {
        "className": "withdrawal",
        "id": "0x963e0000011977",
      },
      "pathFromRoot": Array [
        "b791d53b-d36d-4463-a93b-d6f88836fa7b",
        "bcc61160-44ce-4214-88a0-7beca55cff90",
      ],
      "type": "ECInstanceNode",
    },
    "label": "Baby reboot Pre-emptive",
  },
  Object {
    "description": "Sint sunt ut.",
    "hasChildren": true,
    "imageId": "Kip",
    "isCheckboxEnabled": true,
    "isCheckboxVisible": false,
    "isChecked": false,
    "isEditable": true,
    "isExpanded": true,
    "isSelectionDisabled": true,
    "key": Object {
      "instanceKey": Object {
        "className": "Manager",
        "id": "0x1489c0000004432",
      },
      "pathFromRoot": Array [
        "969ce80c-c7a4-46b3-b4ba-8b1b8f5df1f0",
        "18740c50-8a89-4eb7-ab84-4fa51f5b070a",
      ],
      "type": "ECInstanceNode",
    },
    "label": "Fresh Refined Plastic Chair Gorgeous Granite Chicken",
  },
]
`;

exports[`Node toJSON serializes Node 1`] = `
Object {
  "backColor": undefined,
  "description": "Dicta ut est autem suscipit laboriosam dolore placeat quam.",
  "foreColor": undefined,
  "hasChildren": true,
  "imageId": "array",
  "isCheckboxEnabled": true,
  "isCheckboxVisible": false,
  "isChecked": true,
  "isEditable": false,
  "isExpanded": true,
  "isSelectionDisabled": true,
  "key": Object {
    "instanceKey": Object {
      "className": "Accounts",
      "id": "0x16f0900000018e7",
    },
    "pathFromRoot": Array [
      "49751a56-5950-4ae3-be3b-c39c1d6f2a17",
      "055caccb-47fb-48ce-938e-485a20877ad8",
    ],
    "type": "ECInstanceNode",
  },
  "label": "circuit Dynamic multimedia",
}
`;
